lib/utils/findPort.js

```diff
@@ -1,39 +0,0 @@
-'use strict';
-
-const pRetry = require('p-retry');
-const portfinder = require('portfinder');
-const defaultPort = require('./defaultPort');
-const defaultTo = require('./defaultTo');
-const tryParseInt = require('./tryParseInt');
-
-function runPortFinder() {
-  return new Promise((resolve, reject) => {
-    portfinder.basePort = defaultPort;
-    portfinder.getPort((error, port) => {
-      if (error) {
-        return reject(error);
-      }
-
-      return resolve(port);
-    });
-  });
-}
-
-function findPort(port) {
-  if (port) {
-    return Promise.resolve(port);
-  }
-
-  // Try to find unused port and listen on it for 3 times,
-  // if port is not specified in options.
-  // Because NaN == null is false, defaultTo fails if parseInt returns NaN
-  // so the tryParseInt function is introduced to handle NaN
-  const defaultPortRetry = defaultTo(
-    tryParseInt(process.env.DEFAULT_PORT_RETRY),
-    3
-  );
-
-  return pRetry(runPortFinder, { retries: defaultPortRetry });
-}
-
-module.exports = findPort;
```

Comment:
We should not remove this logic and it is very popular, let's move inside server class, when developers doesn't provide port, we try to use `8080` and if it is busy we try to get next (just reuse this logic)

Reply:
webpack-cli should don't know about internal logic + when we will refactor it to plugin it will be impossible, we call https://github.com/webpack/webpack-cli/blob/master/packages/serve/src/startDevServer.ts#L48 and we can put logic in listen, in think logic will work only when options.port is undefined

I will fix problem here https://github.com/webpack/webpack-cli/blob/master/packages/serve/src/startDevServer.ts#L36, after we do release v4

Reply:
Currently, the caller is responsible for setting the port:
```js
const server = new Server(compiler, {port});
server.listen(port, hostname);
```

Reply:
hm, maybe we should refactor API? and allow to set options only on `Server` constructor?

Reply:
I think it should be the other way around, as the only place the port option is used before `listen()` is creating the socket URL: https://github.com/webpack/webpack-dev-server/blob/1aa0d2db699d3d9edad7d354b4501d2773aa3675/lib/utils/DevServerPlugin.js#L35-L37

I think it should use the `location.port` by default if `sock.port` is not specified (the current `sockPort: 'location'` behavior). I wonder the following test case is valid: https://github.com/webpack/webpack-dev-server/blob/1aa0d2db699d3d9edad7d354b4501d2773aa3675/test/e2e/ClientOptions.test.js#L65-L79 as if one access the page via proxy on `http://localhost:${port2}/main`, should the socket URL also be via proxy on `http://localhost:${port2}/ws`?

Reply:
hm, I think you are right, should we fix it in this PR? maybe we can do it separately?

Reply:
I've opened another PR #2850 since it's a breaking change.

Reply:
Now the problem is we have three sources of ports to use:
(1) `options.port` (not used currently)
(2) an argument passed to `listen`
(3) free port found by `findPort`

Reply:
Any ideas how to better solve it? As I said before we will refactor it to plugin, so plugin will have only `options`, so maybe we should accept port only by options to `Server` class?

Reply:
I think we can remove or deprecate passing the port to `listen`.

Reply:
Maybe we can keep compatibility with warning? if `port` are different, because in the next major release we refactor many places and I think there is no problem with this?

Reply:
https://github.com/webpack/webpack-dev-server/blob/0ab7237e9a2ddbc15db64ade9c0eb3be8c71b2b6/lib/Server.js#L600-L606